pkg/git/mock.go

```diff
@@ -0,0 +1,137 @@
+//
+// Copyright 2023 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package git
+
+import (
+	"fmt"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+type MockGitUrl struct {
+	Protocol string // URL scheme
+	Host     string // URL domain name
+	Owner    string // name of the repo owner
+	Repo     string // name of the repo
+	Revision string // branch name, tag name, or commit id
+	Path     string // path to a directory or file in the repo
+	token    string // used for authenticating a private repo
+	IsFile   bool   // defines if the URL points to a file in the repo
+}
+
+func (m *MockGitUrl) GetToken() string {
+	return m.token
+}
+
+var mockExecute = func(baseDir string, cmd CommandType, args ...string) ([]byte, error) {
+	if cmd == GitCommand {
+		if len(args) > 0 && args[0] == "clone" {
+			u, _ := url.Parse(args[1])
+			password, hasPassword := u.User.Password()
+
+			// private repository
+			if hasPassword {
+				switch password {
+				case "valid-token":
+					_, err := os.Create(filepath.Clean(baseDir) + "/private-repo-resource")
+					if err != nil {
+						return nil, fmt.Errorf("failed to create test resource: %v", err)
+					}
+					return []byte(""), nil
+				default:
+					return []byte(""), fmt.Errorf("not a valid token")
+				}
+			}
+			// public repository
+			_, err := os.Create(filepath.Clean(baseDir) + "/public-repo-resource")
+			if err != nil {
+				return nil, fmt.Errorf("failed to create test resource: %v", err)
+			}
+			return []byte(""), nil
+		}
+
+		if len(args) > 0 && args[0] == "switch" {
+			revision := strings.TrimPrefix(args[2], "origin/")
+			if revision != "invalid-revision" {
+				_, err := os.Create(filepath.Clean(baseDir) + "/valid-git-switch")
```

Comment:
When we do a revision switch, you can have the same resource but different file contents which is why I suggested to write out the byte contents into a file and then check if the file exists with the right content.